src/script/interpreter.cpp

```diff
@@ -1680,6 +1712,100 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                            valtype vchPrevVout;
+                            auto vout_le = htole32(inp.prevout.n);
+                            vchPrevVout.insert(vchPrevVout.begin(), (unsigned char*)&vout_le, (unsigned char*)&vout_le + 4);
+                            stack.push_back(vchPrevTxid);
+                            stack.push_back(vchPrevVout);
+                            break;
+                        }
+                        case 2:
+                        {
+                            pushasset(stack, spent_utxo.nAsset);
+                            break;
+                        }
+                        case 3:
+                        {
+                            pushvalue(stack, spent_utxo.nValue);
+                            break;
+                        }
+                        case 4:
+                        {
+                            valtype vchScriptPubKey;
+                            vchScriptPubKey.insert(vchScriptPubKey.begin(), spent_utxo.scriptPubKey.begin(), spent_utxo.scriptPubKey.end());
+                            stack.push_back(vchScriptPubKey);
+                            break;
+                        }
+                        case 5:
+                        {
+                            valtype vchnSequence;
+                            auto nsequence_le = htole32(inp.nSequence);
+                            vchnSequence.insert(vchnSequence.begin(), (unsigned char*)&nsequence_le, (unsigned char*)&nsequence_le + 4);
+                            stack.push_back(vchnSequence);
+                            break;
+                        }
+                        case 6:
+                        {
+                            if (!inp.assetIssuance.IsNull()) {
+                                valtype vchAssetBlindingNonce, vchAssetEntropy;
+                                pushvalue(stack, inp.assetIssuance.nInflationKeys);
+                                pushvalue(stack, inp.assetIssuance.nAmount);
+                                // Next push Asset entropy
+                                vchAssetEntropy.insert(vchAssetEntropy.begin(), inp.assetIssuance.assetEntropy.begin(), inp.assetIssuance.assetEntropy.end());
+                                stack.push_back(vchAssetEntropy);
+                                // Finally push blinding nonce
+                                // By pushing the this order, we make sure that the stack top is empty
+                                // iff there is no issuance. Note that nInflationKeys can be null and can push false
+                                vchAssetBlindingNonce.insert(vchAssetBlindingNonce.begin(), inp.assetIssuance.assetBlindingNonce.begin(), inp.assetIssuance.assetBlindingNonce.end());
+                                stack.push_back(vchAssetBlindingNonce);
+                            } else { // No issuance
+                                stack.push_back(vchFalse);
+                            }
```

Comment:
It occurs to me that there is no way to get the actual asset ID (though maybe you can gin it up using the streaming hash opcodes..). Worth thinking about in a future softfork, I don't want to extend the scope of this one too much.

Reply:
I think we can compute the asset ID with our CAT and SHA256.  If not, we can add new opcodes at a later date.